core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -666,6 +666,29 @@ static class JoinConditionBasedPredicateInference {
       equivalence = BitSets.closure(equivalence);
     }
 
+    /**
```

Comment:
An optional suggestion for the comment:

> As RexPermuteInputsShuttle, with the modification that an RexInputRef is updated with the type the type found in the input fields (instead of keeping the original type).
> This is use when generating the Left/RightInferredPredicates, permitting different but compatible types for expressions which are moved around the tree (such as moving an expression to a location where a an input type's nullability is different).